docs/ErrorHandlingRationale.md

```diff
@@ -0,0 +1,1940 @@
+    zero value.
+
+The clang attribute would specify how to test the return value for an
+error. For example:
+
+```objc
++ (NSInteger)writePropertyList:(id)plist
+                      toStream:(NSOutputStream *)stream
+                        format:(NSPropertyListFormat)format
+                       options:(NSPropertyListWriteOptions)opt
+                         error:(out NSError **)error
+    NS_ERROR_RESULT(0)
+
+- (AVKeyValueStatus)statusOfValueForKey:(NSString *)key
+                                  error:(NSError **)
+    NS_ERROR_RESULT(AVKeyValueStatusFailed);
+```
+
+We should also provide a Clang attribute which specifies that the
+correct way to test for an error is to check the out-parameter. Both of
+these attributes could potentially be used by the static analyzer, not
+just Swift. (For example, they could try to detect an invalid error
+check.)
+
+A constant value would be sufficient for the cases I've seen, but if the
+argument has to generalized to a simple expression, that's still
+feasible.
+
+##### The error parameter
+
+The obvious import rule for Objective-C methods with `NSError**`
+out-parameters is to simply mark them `throws` and remove the selector
+clause corresponding to the out-parameter. That is, a method like this
+one from `NSAttributedString`:
+
+```objc
+- (NSData *)dataFromRange:(NSRange)range
+       documentAttributes:(NSDictionary *)dict
+                    error:(NSError **)error;
+```
+
+would be imported as:
+
+```swift
+func dataFromRange(
+    _ range: NSRange,
+    documentAttributes dict: NSDictionary
+) throws -> NSData
+```
+
+However, applying this rule haphazardly causes problems for Objective-C
+interoperation, because multiple methods can be imported the same way.
+The model is far more comprehensible to both compiler and programmer if
+the original Objective-C declaration can be unambiguously reconstructed
+from a Swift declaration.
+
+There are two sources of this ambiguity:
+
+-   The error parameter could have appeared at an arbitrary position in
+    the selector; that is, both `foo:bar:error:` and `foo:error:bar:`
+    would appear as `foo:bar:` after import.
+-   The error parameter could have had an arbitrary selector chunk; that
+    is, both `foo:error:` and `foo:withError:` would appear as `foo:`
+    after import.
+
+To allow reconstruction, then, we should only apply the rule when the
+error parameter is the last parameter and the corresponding selector is
+either `error:` or the first chunk. Empirically, this seems to do the
+right thing for all but two sets of APIs in the public API:
+
+-   The `ISyncSessionDriverDelegate` category on `NSObject` declares
+    half-a-dozen methods like this:
+
+```objc
+- (BOOL)sessionDriver:(ISyncSessionDriver *)sender
+    didRegisterClientAndReturnError:(NSError **)outError;
+```
+
+    Fortunately, these delegate methods were all deprecated in Lion, and
+    Swift currently doesn't even import deprecated methods.
+
```

Comment:
Underindented code block.

Reply:
I had indented code blocks in one of the previous PRs, and I think you said that those don't render correctly, didn't you?

Reply:
Yeah, but those were over-indented relative to the text. This is the opposite case — the code block belongs to the bullet. Check out how it renders.